marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs

```diff
@@ -0,0 +1,564 @@
+import Ouroboros.Network.Point (WithOrigin (..))
+
+-- | PostgreSQL implementation for the chain sync database queries.
+databaseQueries :: DatabaseQueries Session
+databaseQueries = DatabaseQueries
+ commitRollback
+ commitBlocks
+ commitGenesisBlock
+ getHeaderAtPoint
+ getIntersectionPoints
+ getGenesisBlock
+
+-- GetGenesisBlock
+
+getGenesisBlock :: GetGenesisBlock Session
+getGenesisBlock = GetGenesisBlock $ statement () $ refineResult (decodeResults . V.toList)
+  [vectorStatement|
+    SELECT block.id :: bytea
+         , tx.id :: bytea
+         , txOut.lovelace :: bigint
+         , txOut.address :: bytea
+    FROM   chain.block AS block
+    JOIN   chain.tx    AS tx    ON tx.blockId = block.id
+                               AND tx.slotNo = block.slotNo
+    JOIN   chain.txOut AS txOut ON txOut.txId   = tx.id
+    WHERE  block.slotNo = -1
+  |]
+  where
+    decodeResults :: [(ByteString, ByteString, Int64, ByteString)] -> Either Text (Maybe GenesisBlock)
+    decodeResults [] = Right Nothing
+    decodeResults rows@((hash, _, _, _) : _) =
+      Just . GenesisBlock (HeaderHash $ toShort hash) . Set.fromList <$> traverse decodeGenesisTx rows
+
+    decodeGenesisTx :: (ByteString, ByteString, Int64, ByteString) -> Either Text GenesisTx
+    decodeGenesisTx (_, txId, lovelace, address) = GenesisTx
+      <$> decodeTxId txId
+      <*> pure (fromIntegral lovelace)
+      <*> decodeAddressAny address
+
+-- GetHeaderAtPoint
+
+getHeaderAtPoint :: GetHeaderAtPoint Session
+getHeaderAtPoint = GetHeaderAtPoint \case
+  ChainPointAtGenesis -> pure Origin
+  point@(ChainPoint slotNo hash) ->
+    statement (slotNoToParam slotNo, headerHashToParam hash) $ refineResult decodeResults
+      [maybeStatement|
+        SELECT block.blockNo :: bigint
+          FROM chain.block AS block
+        WHERE block.slotNo = $1 :: bigint
+          AND block.id     = $2 :: bytea
+      |]
+    where
+      decodeResults :: Maybe Int64 -> Either Text (WithOrigin BlockHeader)
+      decodeResults Nothing        = Left $ "No block found at " <> T.pack (show point)
+      decodeResults (Just blockNo) = Right $ At $ BlockHeader slotNo hash $ BlockNo $ fromIntegral blockNo
+
+decodeTxId :: ByteString -> Either Text TxId
+decodeTxId txId = case deserialiseFromRawBytes AsTxId txId of
+  Nothing    -> Left "Invalid TxId"
+  Just txId' -> Right txId'
+
+decodeAddressAny :: ByteString -> Either Text AddressAny
+decodeAddressAny address = case deserialiseFromRawBytes AsAddressAny address of
+  Nothing       -> Left "Invalid address"
+  Just address' -> Right address'
+
+-- GetIntersectionPoints
+
+getIntersectionPoints :: GetIntersectionPoints Session
+getIntersectionPoints = GetIntersectionPoints $ statement () $ rmap decodeResults
+  [vectorStatement|
+    SELECT slotNo :: bigint, id :: bytea
+    FROM   chain.block
+    WHERE  rollbackToBlock IS NULL
+    ORDER BY slotNo DESC LIMIT 2160
```

Comment:
The Ouroboros security parameter is 2160 _blocks_, not 2160 _slots_, so this limit is not large enough. Also, I don't know that _k_ will ever change, but maybe we should flag this hardcoding so we can use a parameter someday?

Reply:
Yeah, I want to parameterize this, but do not know the most appropriate place to read it from. And this does limit by 2160 blocks, not slots, as it is a `LIMIT`on the `block` table.

Reply:
Sorry, my mistake--I should have realized that when I read the query!

BTW, some of the testnets have a non-standard security parameter.